core/src/main/java/com/linecorp/armeria/common/logging/DefaultRequestLog.java

```diff
@@ -177,192 +166,313 @@ public DefaultRequestLog(RequestContext ctx, ContentPreviewerFactory requestCont
+    @Override
+    public boolean isRequestComplete() {
+        return isAvailable(flags, RequestLogProperty.FLAGS_REQUEST_COMPLETE);
     }
 
     @Override
-    public void endResponseWithLastChild() {
-        checkState(!hasLastChild, "last child is already added");
-        checkState(children != null && !children.isEmpty(), "at least one child should be already added");
-        hasLastChild = true;
-        final RequestLog lastChild = children.get(children.size() - 1);
-        propagateResponseSideLog(lastChild);
+    public boolean isAvailable(RequestLogProperty property) {
+        return isAvailable(flags, property.flag());
     }
 
-    private void propagateResponseSideLog(RequestLog lastChild) {
-        // update the available logs if the lastChild already has them
-        if (lastChild.isAvailable(RESPONSE_START)) {
-            startResponse0(lastChild.responseStartTimeNanos(), lastChild.responseStartTimeMicros(), true);
-        }
+    private boolean isAvailable(int interestedFlags) {
+        return isAvailable(flags, interestedFlags);
+    }
 
-        if (lastChild.isAvailable(RESPONSE_FIRST_BYTES_TRANSFERRED)) {
-            responseFirstBytesTransferred(lastChild.responseFirstBytesTransferredTimeNanos());
-        }
+    private static boolean isAvailable(int flags, RequestLogProperty property) {
+        return isAvailable(flags, property.flag());
+    }
 
-        if (lastChild.isAvailable(RESPONSE_HEADERS)) {
-            responseHeaders(lastChild.responseHeaders());
+    private static boolean isAvailable(int flags, int interestedFlags) {
+        return (flags & interestedFlags) == interestedFlags;
+    }
+
+    @Override
+    public RequestLog partial() {
+        return flags != RequestLogProperty.FLAGS_ALL_COMPLETE ? this : notCheckingAccessor;
+    }
+
+    @Override
+    public CompletableFuture<RequestLog> completeFuture() {
+        return future(RequestLogProperty.FLAGS_ALL_COMPLETE);
+    }
+
+    private <T extends RequestOnlyLog> CompletableFuture<T> future(int interestedFlags) {
+        if (interestedFlags == 0) {
+            throw new IllegalArgumentException("no availability specified");
         }
 
-        if (lastChild.isAvailable(RESPONSE_CONTENT)) {
-            responseContent(lastChild.responseContent(), lastChild.rawResponseContent());
+        final RequestLogFuture newFuture;
+        synchronized (futures) {
+            final RequestLogFuture oldFuture = futures.get(interestedFlags);
+            if (oldFuture != null) {
+                @SuppressWarnings("unchecked")
+                final CompletableFuture<T> cast = (CompletableFuture<T>) oldFuture;
+                return cast;
+            }
+
+            newFuture = new RequestLogFuture(interestedFlags);
+            futures.put(interestedFlags, newFuture);
         }
 
-        if (lastChild.isAvailable(RESPONSE_END)) {
-            propagateResponseEndData(lastChild);
+        if (isAvailable(flags, interestedFlags)) {
```

Comment:
I guess it might be a tiny bit faster to special case so only one volatile read

```java
final int flags = this.flags;
if (flags == availableFlags && flags == FLAGS_REQUEST_COMPLETE) {
  newFuture.completeLog(notCheckingAccessor);
} else if (isAvailable(flags, interestedFlags)) {
  newFuture.completeLog(this);
}
```

Reply:
Fixed slightly differently, but basically one less volatile read. :-)